jina/peapods/networking.py

```diff
@@ -104,67 +93,207 @@ def has_connection(self, address: str) -> bool:
+                        random.randrange(0, len(self._pods[pod][type]))
+                    ]
+                else:
+                    return self._pods[pod][type][entity_id]
+            except KeyError:
+                if (
+                    entity_id is None
+                    and pod in self._pods
+                    and len(self._pods[pod][type])
+                ):
+                    # This can happen as a race condition when removing connections while accessing it
+                    # In this case we dont care for the concrete entity, so retry with the first one
+                    return self._get_connection_list(pod, type, 0)
+                return None
+
+        def _add_pod(self, pod: str):
+            if pod not in self._pods:
+                self._pods[pod] = {'shards': {}, 'heads': {}}
+
+        def _add_connection(
+            self,
+            pod: str,
+            entity_id: int,
+            address: str,
+            connection: Callable[[str], jina_pb2_grpc.JinaDataRequestRPCStub],
+            type: str,
+        ):
+            self._add_pod(pod)
+            if entity_id not in self._pods[pod][type]:
+                connection_list = ReplicaList()
+                self._pods[pod][type][entity_id] = connection_list
+
+            if not self._pods[pod][type][entity_id].has_connection(address):
+                self._logger.debug(
+                    f'Adding connection for pod {pod}/{type}/{entity_id} to {address}'
+                )
+                self._pods[pod][type][entity_id].add_connection(
+                    address, connection(address)
+                )
+
+        def remove_head(self, pod, address, head_id: Optional[int] = 0):
+            return self._remove_connection(pod, head_id, address, 'heads')
+
+        def remove_replica(self, pod, address, shard_id: Optional[int] = 0):
+            return self._remove_connection(pod, shard_id, address, 'shards')
 
+        def _remove_connection(self, pod, entity_id, address, type):
+            if pod in self._pods and entity_id in self._pods[pod][type]:
+                self._logger.debug(
+                    f'Removing connection for pod {pod}/{type}/{entity_id} to {address}'
+                )
+                connection = self._pods[pod][type][entity_id].remove_connection(address)
+                if not self._pods[pod][type][entity_id].has_connections():
+                    del self._pods[pod][type][entity_id]
+                return connection
+            return None
+
+    def __init__(self, logger: Optional[JinaLogger] = None):
         self._logger = logger or JinaLogger(self.__class__.__name__)
+        self._connections = self._ConnectionPoolMap(self._logger)
 
     def send_message(
         self,
         msg: Message,
-        target_address: str,
+        pod: str,
+        head: bool,
+        shard_id: Optional[int] = None,
         polling_type: PollingType = PollingType.ANY,
```

Comment:
Why do we have polling type here? Is the connection pool responsible for this? Should it not be handled at `shard` level?

Reply:
In the current implementation I do the polling on the connection pool level. Of course we could in theory also do this in the runtime, but I think its fits better here? Especially if want to have eventually things like polling ROLLING or so and go over replicas etc. The runtime may have not much knowledge about those

Reply:
Ok